src/core/Microtuner.cpp

```diff
@@ -0,0 +1,253 @@
+	}
+	else
+	{
+		return m_instrumentTrack->firstKeyModel()->value();
+	}
+}
+
+
+/** \brief Return last mapped key, based on currently selected keymap or user selection.
+ *  \return Number ranging from 0 to NumKeys -1
+ */
+int Microtuner::lastKey() const
+{
+	if (enabled() && keyRangeImport())
+	{
+		return Engine::getSong()->getKeymap(m_keymapModel.value())->getLastKey();
+	}
+	else
+	{
+		return m_instrumentTrack->lastKeyModel()->value();
+	}
+}
+
+
+/** \brief Return base key number, based on currently selected keymap or user selection.
+ *  \return Number ranging from 0 to NumKeys -1
+ */
+int Microtuner::baseKey() const
+{
+	if (enabled() && keyRangeImport())
+	{
+		return Engine::getSong()->getKeymap(m_keymapModel.value())->getBaseKey();
+	}
+	else
+	{
+		return m_instrumentTrack->baseNoteModel()->value();
+	}
+}
+
+
+/** \brief Return frequency assigned to the base key, based on currently selected keymap.
+ *  \return Frequency in Hz
+ */
+float Microtuner::baseFreq() const
+{
+	if (enabled())
+	{
+		return Engine::getSong()->getKeymap(m_keymapModel.value())->getBaseFreq();
+	}
+	else
+	{
+		return DefaultBaseFreq;
+	}
+}
+
+
+/** \brief Return frequency for a given MIDI key, using the active mapping and scale.
+ *  \param key A MIDI key number ranging from 0 to 127.
+ *  \param detune Note detuning amount (+- cents).
+ *  \return Frequency in Hz; 0 if key is out of range or not mapped.
+ */
+float Microtuner::keyToFreq(int key, float detune) const
+{
+	if (key < firstKey() || key > lastKey()) {return 0;}
+	Song *song = Engine::getSong();
+	if (!song) {return 0;}
+
+	// Get keymap and scale selected at this moment
+	const unsigned int keymap_id = m_keymapModel.value();
+	const unsigned int scale_id = m_scaleModel.value();
+	std::shared_ptr<const Keymap> keymap = song->getKeymap(keymap_id);
+	std::shared_ptr<const Scale> scale = song->getScale(scale_id);
+	const std::vector<Interval> &intervals = scale->getIntervals();
+
+	// Convert MIDI key to scale degree + octave offset.
+	// The octaves are primarily driven by the keymap wraparound: octave count is increased or decreased if the key
+	// goes over or under keymap range. In case the keymap refers to a degree that does not exist in the scale, it is
+	// assumed the keymap is non-repeating or just really big, so the octaves are also driven by the scale wraparound.
+	const int keymapDegree = keymap->getDegree(key);		// which interval should be used according to the keymap
+	if (keymapDegree == -1) {return 0;}						// key is not mapped, abort
+	const int keymapOctave = keymap->getOctave(key);		// how many times did the keymap repeat
+	const int octaveDegree = intervals.size() - 1;			// index of the interval with octave ratio
+	if (octaveDegree == 0) {								// octave interval is 1/1, i.e. constant base frequency
+		return keymap->getBaseFreq() * powf(2.f, detune / (100 * 12.f));	// → return detuned baseFreq directly
```

Comment:
I don't quite understand what you mean (and why removing it would be OK), but I could (and will) certainly de-duplicate the two occurrences in the function, it will make it easier to maintain and the lines shorter.

Reply:
Oh, now I see, you meant to delete it in combination with the following suggestion. It originally made sense to let Microtuner handle the detuning, since my intention was to have detuning work differently with "exotic" scales. It turned out to be problematic, so I reverted to using standard cents. So removing it from the function and saving an extra call makes sense now, thanks.